hydroflow_lang/src/graph/serde_graph.rs

```diff
@@ -0,0 +1,174 @@
+                        write,
+                        r#"        {id:?}["{id:?} <tt>{code}</tt>"]"#,
+                        id = node_id.data(),
+                        code = self
+                            .nodes
+                            .get(node_id)
+                            .unwrap()
+                            .clone()
+                            .replace('&', "&amp;")
+                            .replace('<', "&lt;")
+                            .replace('>', "&gt;")
+                            .replace('"', "&quot;")
+                            .replace('\n', "<br>"),
+                    )?;
+                }
+            }
+            writeln!(write, "    end")?;
+        }
+        writeln!(write)?;
+        for (node_id, _) in self.nodes.iter() {
+            if self.handoffs.contains_key(node_id) {
+                writeln!(write, r#"    {:?}{{"handoff"}}"#, node_id.data())?;
+            }
+        }
+        writeln!(write)?;
+        for (src, dests) in self.edges.iter() {
+            for dst in dests {
+                writeln!(write, "    {:?}-->{:?}", src.data(), dst.data())?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn serde_to_dot(&self) -> String {
+        let mut output = String::new();
+        self.serde_write_dot(&mut output).unwrap();
+        output
+    }
+
+    pub fn serde_write_dot_edge(
+        &self,
+        src: GraphNodeId,
+        dst: GraphNodeId,
+        tab: usize,
+        w: &mut impl std::fmt::Write,
+    ) -> std::fmt::Result {
+        writeln!(
+            w,
+            "{:t$}{} -> {}",
+            "",
+            src.data().as_ffi(),
+            dst.data().as_ffi(),
+            t = tab,
+        )?;
+        Ok(())
+    }
+    pub fn serde_write_dot(&self, w: &mut impl std::fmt::Write) -> std::fmt::Result {
+        writeln!(w, "digraph {{")?;
+        let mut tab: usize = 2;
+
+        // write out nodes
+        writeln!(w, "{:t$}{{", "", t = tab)?;
+        tab += 2;
+        writeln!(w, "{:t$}node [shape=box]", "", t = tab)?;
+        for (node_id, text) in self.nodes.iter() {
+            let nm = text.replace('"', "\\\"").replace('\n', "\\l");
+            let label = format!("{}", node_id.data().as_ffi());
+            write!(
+                w,
+                "{:t$}{} [label=\"{}{}\"",
+                "",
+                label,
+                nm.clone(),
+                match nm.contains("\\l") {
+                    // if contains linebreak left-justify by appending another "\\l"
+                    true => "\\l",
+                    false => "",
+                },
+                t = tab
+            )?;
+            if self.handoffs.contains_key(node_id) {
+                write!(w, ", shape=diamond")?;
+            }
+            writeln!(w, "]")?;
+        }
+        tab -= 2;
+        writeln!(w, "{:t$}}}", "", t = tab)?;
+
+        // write out edges per subgraph
+        for (sg_id, nodes) in &self.subgraph_nodes {
```

Comment:
Prefer `for (sg_id, nodes) in self.subgraph_nodes.iter()` to match the rest of the code